es6/animations.js

```diff
@@ -5,180 +5,176 @@ import {
+            return [delta, start + delta]
+        } else if (end !== undefined && delta === undefined) {
+            return [end - start, end]
+        } else if (start + delta !== end) {
+            console.log((new Error()).stack)
+            throw `start (${start}) + delta (${delta}) !== end (${end})`
         }
+    }
+    return [delta, end]
+}
 
-        return {start_state, end_state, amt}
-    } else {
-        if (start_state === undefined) start_state = {}
-        if (end_state === undefined) end_state = {}
-        if (amt === undefined) amt = {}
+const exactlyOneIsUndefined = (val1, val2) => {
+    return ((val1 === undefined || val2 === undefined)
+            && !(val1 === undefined && val2 === undefined))
+}
 
-        if (key !== undefined)
-            return checked_animation_amt({start_state: start_state[key], end_state: end_state[key], amt: amt[key]})
+const isNumber = (val) => {
+    return typeof(val) === 'number'
+}
 
-        if (typeof(start_state) !== 'object' || typeof(end_state) !== 'object' || typeof(amt) !== 'object') {
-            throw 'Incompatible types passed as {start_state, end_state, amt}, must all be objects or numbers'
-        }
+const validateAnimation = (animation) => {
+    const end_time = animation.end_time
+    const end_state = animation.end_state
 
-        let keys = Object.keys(start_state)
-        if (!keys.length) keys = Object.keys(end_state)
-        if (!keys.length) keys = Object.keys(amt)
-
-        return keys.reduce((acc, key) => {
-            const amts = checked_animation_amt({start_state: start_state[key], end_state: end_state[key], amt: amt[key]})
-            acc.start_state[key] = amts.start_state
-            acc.end_state[key] = amts.end_state
-            acc.amt[key] = amts.amt
-            return acc
-        }, {start_state: {}, end_state: {}, amt: {}})
+    const computed_end_state = computeAnimatedState([animation], _end_time)
+    if (!isEqual(computed_end_state, end_state)) {
+        throw `Invalid Animate: end_state !== computed_end_state for animation:`
+            + `\n${JSON.stringify(animation, null, '  ')}:`
+            + `${JSON.stringify(computed_end_state, null, '  ')} !==`
+            + `${JSON.stringify(end_state, null, '  ')}`
     }
 }
 
-const KeyedAnimation = ({type, path, key, start_time, end_time, duration, start_state, end_state, amt, curve, unit}) =>
-    Animate({
-        type,
-        path: `${path}/${key}`,
-        key,
-        start_time,
-        end_time,
-        duration,
-        start_state: start_state && start_state[key],
-        end_state: end_state && end_state[key],
-        amt: amt && amt[key],
-        curve, unit,
-    })
+export const Animate = ({
+        type, path, start_time, end_time, duration, start_state, end_state,
+        delta_state, merge=false, curve='linear', unit=null, tick=null,
+        error_checking=true}) => {
 
+    let _start_time, _end_time, _duration,
```

Comment:
Not a fan of this underscore style, `_` intuitively means unused or internal, not "new value".

Why not just make a new obj instead?

```javascript
const animation = {
    ...
}
animation.start_time = (start_time === undefined) ? Date.now() : start_time
...
```

Reply:
I guess my reasoning is that it's a temporary value, internal to the function?

I don't like declaring an object and then mutating it line-by-line; I think it's nice to see the structure of the output clearly defined at the return statement.